src/vs/workbench/api/common/extHostLanguageModelTools.ts

```diff
@@ -51,13 +76,22 @@ export class ExtHostLanguageModelTools implements ExtHostLanguageModelToolsShape
 			.map(tool => typeConvert.LanguageModelToolDescription.to(tool));
 	}
 
-	async $invokeTool(name: string, parameters: any, token: CancellationToken): Promise<IToolResult> {
-		const item = this._registeredTools.get(name);
+	async $invokeTool(dto: IToolInvokationDto, token: CancellationToken): Promise<IToolResult> {
+		const item = this._registeredTools.get(dto.toolId);
 		if (!item) {
-			throw new Error(`Unknown tool ${name}`);
+			throw new Error(`Unknown tool ${dto.toolId}`);
+		}
+
+		const options: vscode.LanguageModelToolInvokationOptions = { parameters: dto.parameters };
+		if (dto.tokenBudget !== undefined) {
+			options.tokenOptions = {
+				tokenBudget: dto.tokenBudget,
+				countTokens: this._tokenCountFuncs.get(dto.callId) || ((value, token = CancellationToken.None) =>
+					this._proxy.$countTokensForInvokation(dto.callId, typeof value === 'string' ? value : typeConvert.LanguageModelChatMessage.from(value), token))
+			};
 		}
 
-		const extensionResult = await item.tool.invoke(parameters, token);
+		const extensionResult = await item.tool.invoke(options, token);
```

Comment:
unrelated to this PR: we should "race cancellation" here and not just rely on extensions honoring the token.